README.md
# snabbdom-timeline

an interactive timeline component implemented in snabbdom

![alt text](example.png "screenshot")


## features
* can stack multiple graphs
* single time or ranged time selectors
* data oriented, functional design
* tiny! (~ 400 lines of code)
* renders with SVG or canvas


## usage

requires node >= v12.17 or a browser that supports the es module format.


via `npm`:
```javascript
import html     from 'snabby'
import timeline from 'snabbdom-timeline'
```

via a pure es module (browser or `deno`):
```javascript
import html     from 'https://cdn.jsdelivr.net/npm/snabby@2/snabby.js'
import timeline from 'https://cdn.jsdelivr.net/gh/mreinstein/snabbdom-timeline/timeline.js'
```


```javascript

// contains all data needed to render a timeline component

const model = {
    container: document.createElement('div'),
    width: 0,
    renderer: 'svg', // canvas | svg
    graphs: [
        {
            title: 'test title',
            label: 'title',
            type: 'scatterPlot',  // scatterPlot | linePlot
            timeRange: {
                start: 0,  // seconds
                end: 0     // seconds
            },
            yRange: {
                start: 0,
                end: 100
            },

            // optional: render a selection control
            selection: {
                type: 'range',
                start: 0,       // seconds | 0
                end: Infinity,  // seconds | Infinity
                dragging: false
            },

            height: 40,              // pixels
            dataColor: 'dodgerblue', // color of data points on the graph
            renderTicks: false,
            renderValueLabel: false,

            // the data points to render
            data: [ ],

            // optional: settings for grid background lines
            gridLines: {
                vertical: {
                    majorColor: '#dedede',
                    minorColor: '#f3f3f3',
                    ticksPerMinor: 2.5,
                    ticksPerMajor: 10
                },
                horizontal: {
                    color: '#eeeeee',
                    lineCount: 2
                }
            }
        }
    ]
}


document.body.appendChild(model.container)


function update () {
    const oldVnode = model.container
    const newVnode = timeline(model, update)
    model.container = html.update(oldVnode, newVnode)
}


// pump test data into the graph
setInterval(function () {
    const value = model.graphs[0].yRange.start + Math.round(Math.random() * (model.graphs[0].yRange.end - model.graphs[0].yRange.start))
    model.graphs[0].data.push({
        t: performance.now() / 1000,
        value
    })

    update()
}, 2000)

update()

```

## graph types

There are 2 available graph types:

### scatterPlot

![alt text](plot-scatter.png "scatter plot graph")


### linePlot

![alt text](plot-line.png "line plot graph")


You can also open `example.html` in a browser to see a more complicated example with 2 graphs linked together.


## renderer option

by default, snabbdom-timeline will use `svg` to render these graphs. These are nice in that they are semantic elements, and are probably more accesible.

However, if you're rendering a _lot_ of graphs with frequently updating data, you may find the svg based rendering performance is insufficient.

For example, in one of my projects I'm
rendering 50 of these graphs, with new data coming in at 60 frames per second, and storing 1,000 data points per graph. Using virtual dom really bogs things down at this scale,
so in that case I use the `canvas` renderer option which has much better performance, at the cost of blurrier text and less semantically friendly markup.


CHANGELOG.md
# changelog

## 0.8.0
* update snabby dependency
* remove unneeded rollup modules


## 0.7.0
* when time selection is all the way to the right, store the `selection.time` value as Infinity


## 0.6.0
* feature: provide a canvas based renderer (fixes #13)
* feature: optionally render fill area under line plot (fixes #11)
* bugfix: fix error occurring when optional gridlines field is missing from model
* bugfix: prevent start and end time range sliders from overlapping (fixes #3)


## 0.5.0
* improve rendering and dragging performance. fixes #9
* make drag handles easier to use. fixes #5


## 0.4.0

* add line plot graph type


## 0.3.1

* tweak readme


## 0.3.0

* update to snabby v2
* require node v12.17+
* enforce build before npm publish


## 0.2.0

* reduce npm package size by specifying `package.json` `file` field
* clarify both npm and pure es module import use cases
* add changelog
* update `snabby` dep to `1.2`
* update `rollup` dep


## 0.1.0

* initial npm version


timeline-canvas.js
import clamp          from 'https://cdn.jsdelivr.net/gh/mreinstein/math-gap/src/clamp.js'
import findPosOnScale from './find-position-on-scale.js'
import html           from 'https://cdn.skypack.dev/snabby?min'
import lerp           from 'https://cdn.jsdelivr.net/gh/mreinstein/math-gap/src/lerp.js'
import throttle       from 'https://cdn.skypack.dev/lodash.throttle'


const DOT_WIDTH = 4
const FPS = 60


function getTimePeriodData (graph) {
    return graph.data.filter((dataPoint) => {
        // if data is outside of the time range skip it
        if (dataPoint.t < graph.timeRange.start)
            return false

        if (dataPoint.t > graph.timeRange.end)
            return false

        return true
    })
}


function getGraphMetrics (model, graph) {
    let leftMargin = 10
    const rightMargin = 10
    let bottomMargin = graph.renderTicks ? 20 : 0

    if (graph.renderValueLabel && graph.selection.type === 'value')
        bottomMargin += 30

    const graphHeight = graph.height - bottomMargin
    const graphWidth = model.width - leftMargin - rightMargin

    return {
        leftMargin, rightMargin, bottomMargin, graphHeight, graphWidth
    }
}


function verticalGridLinesMinor (model, graph, update) {
    if (!graph.gridLines || !graph.gridLines.vertical)
        return

    const m = getGraphMetrics(model, graph)

    const gridLines = [ ]
    const pixelsPerTick = 6
    const pixelsPerMinorLine = pixelsPerTick * graph.gridLines.vertical.ticksPerMinor

    const strokeWidth = 1 / (window.devicePixelRatio || 1)
    const { ctx } = graph

    ctx.lineWidth = 1 //strokeWidth
    ctx.strokeStyle = graph.gridLines.vertical.minorColor

    ctx.beginPath()

    for (let i=0; i < m.graphWidth; i += pixelsPerMinorLine) {
        const x = m.leftMargin + i + 0.5
        ctx.moveTo(x, 0)
        ctx.lineTo(x, m.graphHeight)
    }

    ctx.stroke()
}


function verticalGridLinesMajor (model, graph, update) {
    if (!graph.gridLines || !graph.gridLines.vertical)
        return

    const m = getGraphMetrics(model, graph)

    const pixelsPerTick = 6
    const pixelsPerMajorLine = pixelsPerTick * graph.gridLines.vertical.ticksPerMajor

    const { ctx } = graph

    const strokeWidth = 1 / (window.devicePixelRatio || 1)
    ctx.lineWidth = strokeWidth
    ctx.strokeStyle = graph.gridLines.vertical.majorColor

    ctx.beginPath()

    for (let i=0; i < m.graphWidth; i += pixelsPerMajorLine) {
        const x = m.leftMargin + i + 0.5
        ctx.moveTo(x, 0)
        ctx.lineTo(x, m.graphHeight)
    }

    ctx.stroke()
}


function gridLines (model, graph, update) {
    if (!graph.gridLines || !graph.gridLines.horizontal)
        return

    const m = getGraphMetrics(model, graph)

    const { ctx } = graph

    const strokeWidth = 1 / (window.devicePixelRatio || 1)
    ctx.lineWidth = 0.5 //strokeWidth
    ctx.strokeStyle = graph.gridLines.horizontal.color

    ctx.beginPath()

    ctx.setLineDash([4, 2])

    const distanceBetweenLines = m.graphHeight / (graph.gridLines.horizontal.lineCount + 1)
    for (let y=distanceBetweenLines; y < m.graphHeight; y += distanceBetweenLines) {
        ctx.moveTo(m.leftMargin + 0.5, y + 0.5)
        ctx.lineTo(m.leftMargin + m.graphWidth + 0.5, y + 0.5)
    }

    ctx.stroke()
    ctx.setLineDash([])
}


function renderLinePlotGraph (model, graph, dotWidth) {
    const tp = getTimePeriodData(graph)
    const m = getGraphMetrics(model, graph)

    const { ctx } = graph
    let lastX, lastY

    if (graph.linePlotAreaColor) {
        const region = new Path2D()

        for (let i=0; i < tp.length; i++) {
            const point = tp[i]

            const startX = findPosOnScale(graph.timeRange.start, graph.timeRange.end, point.t)
            const x = Math.round(startX * (m.graphWidth - dotWidth) + m.leftMargin) + 0.5

            const yLength = graph.yRange.end - graph.yRange.start
            const y = Math.round((1 - (point.value / yLength)) * (m.graphHeight - dotWidth)) + 0.5

            if (i === 0)
                region.moveTo(x, m.graphHeight)

            region.lineTo(x, y)
            lastX = x
        }

        region.lineTo(lastX, m.graphHeight)

        region.closePath()
        ctx.fillStyle = graph.linePlotAreaColor
        ctx.fill(region)
    }

    ctx.beginPath()
    ctx.strokeStyle = graph.dataColor
    ctx.lineWidth = 1

    for (let i=0; i < tp.length; i++) {
        const point = tp[i]

        const startX = findPosOnScale(graph.timeRange.start, graph.timeRange.end, point.t)
        const x = Math.round(startX * (m.graphWidth - dotWidth) + m.leftMargin) + 0.5

        const yLength = graph.yRange.end - graph.yRange.start
        const y = Math.round((1 - (point.value / yLength)) * (m.graphHeight - dotWidth)) + 0.5

        if (i > 0) {
            ctx.moveTo(lastX, lastY)
            ctx.lineTo(x, y)
        }

        lastX = x
        lastY = y
    }

    ctx.stroke()
}


function renderScatterPlotGraph (model, graph, dotWidth) {
    const tp = getTimePeriodData(graph)
    const m = getGraphMetrics(model, graph)

    const { ctx } = graph

    ctx.fillStyle = graph.dataColor

    return tp.map((point) => {
        const startX = findPosOnScale(graph.timeRange.start, graph.timeRange.end, point.t)
        const x = startX * (m.graphWidth - dotWidth) + m.leftMargin

        const yLength = graph.yRange.end - graph.yRange.start
        const y = (1 - (point.value / yLength)) * (m.graphHeight - dotWidth)

        ctx.fillRect(Math.round(x), Math.round(y), dotWidth, dotWidth)
    })
}


function graphComponent (model, graph, update) {

    const dotWidth = 4
    const m = getGraphMetrics(model, graph)

    // TODO: change mouse cursor to ew-resize when hovering over a drag handle

    const _mouseMove = throttle(function (ev) {
        const rect = model.elm.getBoundingClientRect()
        const x = clamp(ev.clientX - rect.left, 0, model.elm.clientWidth) //x position within the element.

        if (graph.selection.dragging === 'time') {
            const pos = clamp((x - m.leftMargin) / m.graphWidth, 0, 1)
            graph.selection.time = (pos === 1) ? Infinity : lerp(graph.timeRange.start, graph.timeRange.end, pos)

        } else if (graph.selection.dragging === 'start') {
            const pos = clamp((x - m.leftMargin) / m.graphWidth, 0, 1)

            graph.selection.start = lerp(graph.timeRange.start, graph.timeRange.end, pos)

            // prevent dragging the start control beyond end
            if (graph.selection.start > graph.selection.end)
                graph.selection.start = graph.selection.end

        } else if (graph.selection.dragging === 'end') {
            const pos = clamp((x - m.leftMargin) / m.graphWidth, 0, 1)

            graph.selection.end = (pos === 1) ? Infinity : lerp(graph.timeRange.start, graph.timeRange.end, pos)

            // prevent dragging the end control beyond start
            if (graph.selection.end < graph.selection.start)
                graph.selection.end = graph.selection.start
        }

        update()
    }, 1000 / FPS)

    const _mouseUp = function (ev) {
        graph.selection.dragging = undefined
        document.removeEventListener('mouseup', _mouseUp)
        document.removeEventListener('mousemove', _mouseMove)
        update()
    }

    const _mouseDown = function (ev) {

        let draggingType  // start | end  | time | undefined

        if (graph.selection?.type === 'range' && ev.offsetY <= 21) {
            // check to see if the offsetX is within the bounds of the start or end drag controls
            const startX = findPosOnScale(graph.timeRange.start, graph.timeRange.end, graph.selection.start)
            const endX = findPosOnScale(graph.timeRange.start, graph.timeRange.end, graph.selection.end)

            const x = startX * (m.graphWidth - dotWidth) + m.leftMargin
            const x2 = endX * (m.graphWidth - dotWidth) + m.leftMargin

            if (Math.abs(ev.offsetX - x) < 10) {
                draggingType = 'start'
            } else if (Math.abs(ev.offsetX - x2) < 10) {
                draggingType = 'end'
            }
        } else if (graph.selection?.type === 'value') {
            const m = getGraphMetrics(model, graph)
            if (ev.offsetY >= m.graphHeight + 10)
                draggingType = 'time'
        }

        graph.selection.dragging = draggingType

        if (draggingType) {
            document.addEventListener('mousemove', _mouseMove, { passive: true })
            document.addEventListener('mouseup', _mouseUp)
            update()
        }
    };

    const _insertHook = function (vnode) {
        model.elm = vnode.elm
        graph.ctx = vnode.elm.getContext('2d')
    }

    if (graph.ctx) {
        graph.ctx.clearRect(0, 0, model.elm.width, model.elm.height)
        verticalGridLinesMinor(model, graph, update)
        verticalGridLinesMajor(model, graph, update)
        gridLines(model, graph, update)


        // draw the bottom line of the graph
        graph.ctx.beginPath()
        graph.ctx.strokeStyle = '#888'
        graph.ctx.lineWidth = 1
        graph.ctx.moveTo(m.leftMargin + 0.5, m.graphHeight - 0.5)
        graph.ctx.lineTo(m.leftMargin + m.graphWidth + 0.5, m.graphHeight - 0.5)
        graph.ctx.stroke()


        if (graph.renderTicks) {
            tickMarksComponent(model, graph, update)
            tickLabelsComponent(model, graph, update)
        }

        graph.ctx.strokeStyle = graph.dataColor
        if (graph.type === 'scatterPlot')
            renderScatterPlotGraph(model, graph, dotWidth)
        else
            renderLinePlotGraph(model, graph, dotWidth)
        timeSelectionComponent(model, graph, update)
        renderLabelComponent(model, graph, update)
    }

    if (!graph.key)
        graph.key = 'u' + Math.floor(Math.random() * 9999999)

    return html`<canvas width="${model.width}"
                        height="${graph.height}"
                        @hook:insert=${_insertHook}
                        @key=${graph.key}
                        style="height: ${graph.height}px; width: 100%; padding-top: 10px; background-color: white; image-rendering: pixelated"
                        @style:cursor=${graph.selection.dragging ? 'ew-resize' : 'inherit' }
                        @on:mousedown=${_mouseDown}></canvas>`
}


function renderLabelComponent (model, graph, update) {
    const { ctx } = graph
    const m = getGraphMetrics(model, graph)
    ctx.fillStyle = 'rgba(0, 0, 0, 0.7)'

    if (graph.renderValueLabel && graph.selection.type === 'value') {
        ctx.textAlign = 'start'
        const t = (graph.selection.time === Infinity ? graph.timeRange.end : graph.selection.time).toFixed(1)

        ctx.fillText(`t: ${t}s`, 2, m.graphHeight + m.bottomMargin - 8)
    }

    ctx.textAlign = 'end'
    ctx.fillText(graph.label, m.graphWidth + m.leftMargin - DOT_WIDTH, 12)
}


function tickMarksComponent (model, graph, update) {

    const m = getGraphMetrics(model, graph)

    const { ctx } = graph

    const constPixelsPerTick = 6

    ctx.lineWidth = 1
    ctx.strokeStyle = '#888'

    ctx.beginPath()

    for (let i=0; i < m.graphWidth; i += constPixelsPerTick) {
        const tickHeight = (i % 60 === 0) ? 8 : 4
        const x = m.leftMargin + i + 0.5
        const height = tickHeight
        ctx.moveTo(x, m.graphHeight)
        ctx.lineTo(x, m.graphHeight + tickHeight)
    }

    ctx.stroke()
}


function tickLabelsComponent (model, graph, update) {
    const m = getGraphMetrics(model, graph)

    const timePeriod = (graph.timeRange.end - graph.timeRange.start)

    const { ctx } = graph

    const constPixelsPerTick = 6

    ctx.font = '10px monospace'
    ctx.textAlign = 'center'
    //ctx.fillStyle = '#888'
    ctx.strokeStyle = '#888'
    //ctx.strokeWidth = 1

    const tickCount = m.graphWidth / constPixelsPerTick
    const secondsPerTick = timePeriod / tickCount
    let lastSecond

    // every 10 ticks, draw the seconds
    for (let i=0; i < m.graphWidth; i += 60) {
        const tickIdx = i / constPixelsPerTick
        const seconds = (graph.timeRange.start + tickIdx * secondsPerTick).toFixed(1)
        if (lastSecond !== seconds) {
            lastSecond = seconds
            ctx.strokeText(seconds, m.leftMargin + i, m.graphHeight + 19)
        }
    }
}


function timeSelectionComponent (model, graph, update) {
    if (graph.selection.type === 'range')
        timeRangeSelectionComponent(model, graph, update)

    if (graph.selection.type === 'value')
        timeValueSelectionComponent(model, graph, update)
}


function timeRangeSelectionComponent (model, graph, update) {
    const m = getGraphMetrics(model, graph)
    const startX = findPosOnScale(graph.timeRange.start, graph.timeRange.end, graph.selection.start)
    const endX = findPosOnScale(graph.timeRange.start, graph.timeRange.end, graph.selection.end)

    const { ctx } = graph

    // draw left and right greyed out graph areas (unselected regions)
    ctx.fillStyle = 'rgba(205,205,205, 0.85)'
    ctx.fillRect(m.leftMargin,
                 0,
                 startX * m.graphWidth,
                 m.graphHeight)


    ctx.fillRect(m.leftMargin + (endX * m.graphWidth),
                 0,
                 m.graphWidth - (m.graphWidth*endX),
                 m.graphHeight)

    const downHandlePath = [
        [ -5, -4],
        [  0, -8],
        [ 10,  0],
        [  0,  8]
    ]

    // left drag handle
    renderDragHandle(ctx, m.leftMargin + startX * m.graphWidth, 12, downHandlePath)

    // right drag handle
    renderDragHandle(ctx,m.leftMargin + (endX * m.graphWidth), 12, downHandlePath)

    ctx.strokeStyle = 'rgb(255,64,129)'
    ctx.lineWidth = 1
    ctx.beginPath()
    // left drag line
    ctx.moveTo(m.leftMargin + startX * m.graphWidth - 0.5, 11)
    ctx.lineTo(m.leftMargin + startX * m.graphWidth - 0.5, m.graphHeight+1)

    // right drag line
    ctx.moveTo(m.leftMargin + (endX * m.graphWidth) - 0.5, 11)
    ctx.lineTo(m.leftMargin + (endX * m.graphWidth) - 0.5, m.graphHeight+1)

    ctx.stroke()
}


function renderDragHandle (ctx, startX, startY, path) {
    const region = new Path2D()

    const currentPoint = [ startX, startY ]
    region.moveTo(currentPoint[0], currentPoint[1])

    for (const p of path) {
        currentPoint[0] += p[0]
        currentPoint[1] += p[1]
        region.lineTo(currentPoint[0], currentPoint[1])
    }

    region.closePath()
    ctx.fillStyle = 'rgba(255,64,129)'
    ctx.fill(region)
}


function timeValueSelectionComponent (model, graph, update) {
    const m = getGraphMetrics(model, graph)

    const startX = findPosOnScale(graph.timeRange.start, graph.timeRange.end, graph.selection.time)

    const x = startX * m.graphWidth + m.leftMargin
    const y = m.graphHeight

    const { ctx } = graph

    const upHandlePath = [
        [  5, 4],
        [  0, 8],
        [-10, 0],
        [  0, -8]
    ]

    renderDragHandle(ctx, x, y-1, upHandlePath)

    ctx.strokeStyle = 'rgb(255,64,129)'
    ctx.lineWidth = 1
    ctx.beginPath()

    // drag line
    ctx.moveTo(x - 0.5, 0)
    ctx.lineTo(x - 0.5, m.graphHeight)
    ctx.stroke()
}


export default function timelineComponent (model, update) {
    const _insertHook = function (vnode) {
        model.container = vnode
    }

    if (model.container)
        model.width = model.container.elm ? model.container.elm.offsetWidth : model.container.offsetWidth

    return html`
        <div class="graph-stack"
             @hook:insert=${_insertHook}
             style="width: 100%; display: grid; grid-template-columns: 1fr; border: ${model.border || 'none'};">
            ${model.graphs.map((g) => graphComponent(model, g, update))}
        </div>`
}


timeline-svg.js
import clamp          from 'https://cdn.jsdelivr.net/gh/mreinstein/math-gap/src/clamp.js'
import findPosOnScale from './find-position-on-scale.js'
import html           from 'https://cdn.skypack.dev/snabby?min'
import lerp           from 'https://cdn.jsdelivr.net/gh/mreinstein/math-gap/src/lerp.js'
import throttle       from 'https://cdn.skypack.dev/lodash.throttle'
import { h }          from 'https://cdn.jsdelivr.net/npm/snabbdom@2.1.0/build/package/h.js'; // helper function for creating vnodes


const FPS = 60


function getTimePeriodData (graph) {
    return graph.data.filter((dataPoint) => {
        // if data is outside of the time range skip it
        if (dataPoint.t < graph.timeRange.start)
            return false

        if (dataPoint.t > graph.timeRange.end)
            return false

        return true
    })
}


function getGraphMetrics (model, graph) {
    let leftMargin = 10
    const rightMargin = 10
    let bottomMargin = graph.renderTicks ? 20 : 0

    if (graph.renderValueLabel && graph.selection.type === 'value')
        bottomMargin += 30

    const graphHeight = graph.height - bottomMargin
    const graphWidth = model.width - leftMargin - rightMargin

    return {
        leftMargin, rightMargin, bottomMargin, graphHeight, graphWidth
    }
}


function verticalGridLinesMinor (model, graph, update) {
    if (!graph.gridLines || !graph.gridLines.vertical)
        return html``

    const m = getGraphMetrics(model, graph)

    const gridLines = [ ]
    const pixelsPerTick = 6
    const pixelsPerMinorLine = pixelsPerTick * graph.gridLines.vertical.ticksPerMinor
    for (let i=0; i < m.graphWidth; i += pixelsPerMinorLine) {
        const x = m.leftMargin + i
        //gridLines.push(html`<line x1="${x}" x2="${x}" y1="0" y2="${m.graphHeight}"/>`)
        gridLines.push(h('line', { attrs: { x1: x, x2: x, y1: 0, y2: m.graphHeight } }))
    }

    const strokeWidth = 1 / (window.devicePixelRatio || 1)
    return html`<g class="grid-minor" style="stroke: ${graph.gridLines.vertical.minorColor}; stroke-width: ${strokeWidth}">${gridLines}</g>`
}


function verticalGridLinesMajor (model, graph, update) {
    if (!graph.gridLines || !graph.gridLines.vertical)
        return html``

    const m = getGraphMetrics(model, graph)

    const gridLines = [ ]
    const pixelsPerTick = 6
    const pixelsPerMajorLine = pixelsPerTick * graph.gridLines.vertical.ticksPerMajor
    for (let i=0; i < m.graphWidth; i += pixelsPerMajorLine) {
        const x = m.leftMargin + i
        //gridLines.push(html`<line x1="${x}" x2="${x}" y1="0" y2="${m.graphHeight}"/>`)
        gridLines.push(h('line', { attrs: { x1: x, x2: x, y1: 0, y2: m.graphHeight } }))
    }

    const strokeWidth = 1 / (window.devicePixelRatio || 1)
    return html`<g class="grid-major" style="stroke: ${graph.gridLines.vertical.majorColor}; stroke-width: ${strokeWidth}">${gridLines}</g>`
}


function gridLines (model, graph, update) {
    if (!graph.gridLines || !graph.gridLines.horizontal)
        return html``

    const m = getGraphMetrics(model, graph)

    const gridLines = [ ]
    const distanceBetweenLines = m.graphHeight / (graph.gridLines.horizontal.lineCount + 1)
    for (let y=distanceBetweenLines; y < m.graphHeight; y += distanceBetweenLines) {
        gridLines.push(html`<line x1="${m.leftMargin}" x2="${m.leftMargin + m.graphWidth}" y1="${y}" y2="${y}"/>`)
    }

    const strokeWidth = 1 / (window.devicePixelRatio || 1)

    return html`<g class="grid-horiz"
                   style="stroke: ${graph.gridLines.horizontal.color}; stroke-width: ${strokeWidth}"
                   stroke-dasharray="4 2">${gridLines}</g>`
}


function renderLinePlotGraph (model, graph, dotWidth) {
    const tp = getTimePeriodData(graph)
    const m = getGraphMetrics(model, graph)

    const lines = [ ]
    let lastX, lastY

    let pathString = ''

    for (let i=0; i < tp.length; i++) {
        const point = tp[i]

        const startX = findPosOnScale(graph.timeRange.start, graph.timeRange.end, point.t)
        const x = startX * (m.graphWidth - dotWidth) + m.leftMargin

        const yLength = graph.yRange.end - graph.yRange.start
        const y = (1 - (point.value / yLength)) * (m.graphHeight - dotWidth)

        if (i > 0) {
            lines.push(h('line', { attrs: { x1: lastX, y1: lastY, x2: x, y2: y } }))
            pathString += `L ${x} ${y}`
        } else {
            pathString = `M ${x} ${m.graphHeight}  L ${x} ${y}`
        }

        lastX = x
        lastY = y
    }

    if (graph.linePlotAreaColor && tp.length) {
        pathString += ` L ${lastX} ${m.graphHeight} Z`
        lines.unshift(h('path', { attrs: { d: pathString, fill: graph.linePlotAreaColor, stroke: 'transparent' } }))
    }

    return lines
}


function renderScatterPlotGraph (model, graph, dotWidth) {
    const tp = getTimePeriodData(graph)
    const m = getGraphMetrics(model, graph)

    return tp.map((point) => {
        const startX = findPosOnScale(graph.timeRange.start, graph.timeRange.end, point.t)
        const x = startX * (m.graphWidth - dotWidth) + m.leftMargin

        const yLength = graph.yRange.end - graph.yRange.start
        const y = (1 - (point.value / yLength)) * (m.graphHeight - dotWidth)

        return h('rect', {
            attrs: {
                x,
                y,
                'data-value': point.value,
                width: dotWidth,
                height: dotWidth
            }
        })
    })
}


function graphComponent (model, graph, update) {

    const dotWidth = 4
    const m = getGraphMetrics(model, graph)

    const _stopDragging = function () {
        graph.selection.dragging = undefined
        update()
    }

    const _insertHook = function (vnode) {
        model.elm = vnode.elm
    }

    return html`
        <svg xmlns="http://www.w3.org/2000/svg"
             class="graph"
             aria-labelledby="title"
             role="img"
             viewBox="0 0 ${model.width} ${graph.height}"
             style="height: ${graph.height}px; width: 100%; padding-top: 10px; background-color: white; font-size: 10px; text-anchor: middle; -moz-user-select: none; -webkit-user-select: none; user-select: none; -webkit-user-drag: none; -khtml-user-drag: none; -moz-user-drag: none; -o-user-drag: none; user-drag: none;"
             @on:mouseup=${_stopDragging}
             @hook:insert=${_insertHook}>
            <title id="title">${graph.title}</title>

            ${verticalGridLinesMinor(model, graph, update)}
            ${verticalGridLinesMajor(model, graph, update)}
            ${gridLines(model, graph, update)}

            <g style="stroke: #888; stroke-dasharray: 0; stroke-width: 1;">
                <line x1="${m.leftMargin}" x2="${m.leftMargin+m.graphWidth}" y1="${m.graphHeight}" y2="${m.graphHeight}" />
                ${tickMarksComponent(model, graph, update)}
                ${tickLabelsComponent(model, graph, update)}
            </g>

            <g class="data"
               style="fill: ${graph.dataColor}; stroke: ${graph.dataColor}; stroke-width: 1;">
               ${graph.type === 'scatterPlot' ? renderScatterPlotGraph(model, graph, dotWidth) : renderLinePlotGraph(model, graph, dotWidth)}
            </g>

            ${timeSelectionComponent(model, graph, update)}

            <text x="${m.graphWidth + m.leftMargin - dotWidth}" y="12" style="fill: rgba(0, 0, 0, 0.7); text-anchor: end; pointer-events: none;">${graph.label}</text>
            ${renderLabelComponent(model, graph, update)}
        </svg>`
}


function renderLabelComponent (model, graph, update) {
    if (graph.renderValueLabel && graph.selection.type === 'value') {
        const m = getGraphMetrics(model, graph)
        const t = (graph.selection.time === Infinity ? graph.timeRange.end : graph.selection.time).toFixed(1)
        return html`<text x="2" y="${m.graphHeight + m.bottomMargin - 8}" style="fill: rgba(0, 0, 0, 0.7); text-anchor: start; pointer-events: none;">t: ${t}s</text>`
    }

    return html``
}


function tickMarksComponent (model, graph, update) {
    const m = getGraphMetrics(model, graph)

    const tickMarks = [ ]

    const constPixelsPerTick = 6

    if (graph.renderTicks) {
        for (let i=0; i < m.graphWidth; i += constPixelsPerTick) {
            const tickHeight = (i % 60 === 0) ? 8 : 4
            const x = m.leftMargin + i
            const height = tickHeight
            tickMarks.push(
                h('line', { attrs: { x1: x, x2: x, y1: m.graphHeight, y2: m.graphHeight + tickHeight } })
            )
        }
    }

    return tickMarks
}


function tickLabelsComponent (model, graph, update) {
    const m = getGraphMetrics(model, graph)

    const timePeriod = (graph.timeRange.end - graph.timeRange.start)

    const tickLabels = [ ]
    const constPixelsPerTick = 6

    if (graph.renderTicks) {
        const tickCount = m.graphWidth / constPixelsPerTick
        const secondsPerTick = timePeriod / tickCount
        let lastSecond

        // every 10 ticks, draw the seconds
        for (let i=0; i < m.graphWidth; i += 60) {
            const tickIdx = i / constPixelsPerTick
            const seconds = (graph.timeRange.start + tickIdx * secondsPerTick).toFixed(1)
            if (lastSecond !== seconds) {
                tickLabels.push({ x: m.leftMargin + i, seconds })
                lastSecond = seconds
            }
        }
    }

    return tickLabels.map((tick) => {
        return html`<text x="${tick.x}" y="${m.graphHeight + 19}">${tick.seconds}</text>`
    })
}


function timeSelectionComponent (model, graph, update) {
    if (graph.selection.type === 'range')
        return timeRangeSelectionComponent(model, graph, update)

    if (graph.selection.type === 'value')
        return timeValueSelectionComponent(model, graph, update)

    return html``
}


function timeRangeSelectionComponent (model, graph, update) {
    const m = getGraphMetrics(model, graph)
    const startX = findPosOnScale(graph.timeRange.start, graph.timeRange.end, graph.selection.start)
    const endX = findPosOnScale(graph.timeRange.start, graph.timeRange.end, graph.selection.end)

    const _mouseMove = throttle(function (ev) {
        const rect = model.elm.getBoundingClientRect()
        const x = clamp(ev.clientX - rect.left, 0, model.elm.clientWidth) //x position within the element.

        if (graph.selection.dragging === 'start') {
            const pos = clamp((x - m.leftMargin) / m.graphWidth, 0, 1)
            graph.selection.start = lerp(graph.timeRange.start, graph.timeRange.end, pos)

            // prevent dragging the start control beyond end
            if (graph.selection.start > graph.selection.end)
                graph.selection.start = graph.selection.end

        } else if (graph.selection.dragging === 'end') {
            const pos = clamp((x - m.leftMargin) / m.graphWidth, 0, 1)
            graph.selection.end = (pos === 1) ? Infinity : lerp(graph.timeRange.start, graph.timeRange.end, pos)

            // prevent dragging the end control beyond start
            if (graph.selection.end < graph.selection.start)
                graph.selection.end = graph.selection.start
        }

        update()
    }, 1000 / FPS)

    const _mouseUp = function () {
        graph.selection.dragging = undefined
        document.removeEventListener('mouseup', _mouseUp)
        document.removeEventListener('mousemove', _mouseMove)
        update()
    }

    // @param String position  start | end
    const _mouseDown = function (position) {
        graph.selection.dragging = position
        document.addEventListener('mousemove', _mouseMove, { passive: true })
        document.addEventListener('mouseup', _mouseUp)
        update()
    }

    return html`
        <g class="time-selection">

            <rect x="${m.leftMargin}" y="0" width="${startX * m.graphWidth}" height="${m.graphHeight}"
                    @on:mousedown=${()=> _mouseDown('start')}
                    style="fill: rgba(205,205,205, 0.85);"/>

            <path d="M ${m.leftMargin + startX * m.graphWidth} 12 l -5 -4 l 0 -8    l 10 0  l 0 8 Z"
                  style="fill: rgb(255,64,129); cursor: ew-resize;"
                  @on:mousedown=${()=> _mouseDown('start')}/>

            <line x1="${m.leftMargin + startX * m.graphWidth}"
                  x2="${m.leftMargin + startX * m.graphWidth}"
                  y1="11"
                  y2="${m.graphHeight+1}" stroke="rgb(255,64,129)"/>


            <rect x="${m.leftMargin + (endX * m.graphWidth)}" y="0" width="${m.graphWidth - (m.graphWidth*endX)}" height="${m.graphHeight}"
                    @on:mousedown=${()=> _mouseDown('end')}
                    style="fill: rgba(205,205,205, 0.85);"/>

            <path d="M ${m.leftMargin + (endX * m.graphWidth)} 12 l -5 -4 l 0 -8    l 10 0  l 0 8 Z"
                  style="fill: rgb(255,64,129); cursor: ew-resize;"
                  @on:mousedown=${()=> _mouseDown('end')}/>

            <line x1="${m.leftMargin + (endX * m.graphWidth)}"
                  x2="${m.leftMargin + (endX * m.graphWidth)}"
                  y1="11"
                  y2="${m.graphHeight+1}" stroke="rgb(255,64,129)"/>
        </g>`
}


function timeValueSelectionComponent (model, graph, update) {

    const _mouseMove = throttle(function (ev) {
        const rect = model.elm.getBoundingClientRect()
        const x = clamp(ev.clientX - rect.left, 0, model.elm.clientWidth) //x position within the element.

        const pos = clamp((x - m.leftMargin) / m.graphWidth, 0, 1)
        graph.selection.time = (pos === 1) ? Infinity : lerp(graph.timeRange.start, graph.timeRange.end, pos)
        update()
    }, 1000 / 60)

    const _mouseUp = function () {
        graph.selection.dragging = undefined
        document.removeEventListener('mouseup', _mouseUp)
        document.removeEventListener('mousemove', _mouseMove)
        update()
    }

    const _mouseDown = function (position) {
        graph.selection.dragging = true
        document.addEventListener('mousemove', _mouseMove, { passive: true })
        document.addEventListener('mouseup', _mouseUp)
        update()
    }

    const m = getGraphMetrics(model, graph)

    const startX = findPosOnScale(graph.timeRange.start, graph.timeRange.end, graph.selection.time)

    const x = startX * m.graphWidth + m.leftMargin
    const y = m.graphHeight

    return html`
        <g class="time-selection" @on:mousedown=${_mouseDown}>
            <line x1="${x}"
                  x2="${x}"
                  y1="0"
                  y2="${m.graphHeight}" stroke="deeppink"/>

            <path d="M ${x} ${y-1} l 5 4 l 0 8    l -10 0  l 0 -8 Z"
                  style="fill: rgb(255,64,129); cursor: ew-resize;"
                  />
        </g>`
}


export default function timelineComponent (model, update) {
    const _insertHook = function (vnode) {
        model.container = vnode
    }

    if (model.container)
        model.width = model.container.elm ? model.container.elm.offsetWidth : model.container.offsetWidth

    return html`
        <div class="graph-stack"
             @hook:insert=${_insertHook}
             style="width: 100%; display: grid; grid-template-columns: 1fr; border: ${model.border || 'none'};">
            ${model.graphs.map((g) => graphComponent(model, g, update))}
        </div>`
}
